lib/block_view.go

```diff
@@ -4420,21 +4420,53 @@ func (bav *UtxoView) ConnectBlock(
 		}
 
 		if includeFeesInBlockReward {
-			// Add the fees from this txn to the total fees. If any overflow occurs
-			// mark the block as invalid and return a rule error. Note that block reward
-			// txns should count as having zero fees.
-			if totalFees > (math.MaxUint64 - currentFees) {
-				return nil, RuleErrorTxnOutputWithInvalidAmount
-			}
-			totalFees += currentFees
+			if txn.TxnMeta.GetTxnType() != TxnTypeAtomicTxnsWrapper {
+				// Compute the BMF given the current fees paid in the block.
+				_, utilityFee = computeBMF(currentFees)
+
+				// Add the fees from this txn to the total fees. If any overflow occurs
+				// mark the block as invalid and return a rule error. Note that block reward
+				// txns should count as having zero fees.
+				if totalFees > (math.MaxUint64 - currentFees) {
+					return nil, RuleErrorTxnOutputWithInvalidAmount
+				}
+				totalFees += currentFees
 
-			// For PoS, the maximum block reward is based on the maximum utility fee.
-			// Add the utility fees to the max utility fees. If any overflow
-			// occurs mark the block as invalid and return a rule error.
-			maxUtilityFee, err = SafeUint64().Add(maxUtilityFee, utilityFee)
-			if err != nil {
-				return nil, errors.Wrapf(RuleErrorPoSBlockRewardWithInvalidAmount,
-					"ConnectBlock: error computing maxUtilityFee: %v", err)
+				// For PoS, the maximum block reward is based on the maximum utility fee.
+				// Add the utility fees to the max utility fees. If any overflow
+				// occurs mark the block as invalid and return a rule error.
+				maxUtilityFee, err = SafeUint64().Add(maxUtilityFee, utilityFee)
+				if err != nil {
+					return nil, errors.Wrapf(RuleErrorPoSBlockRewardWithInvalidAmount,
+						"ConnectBlock: error computing maxUtilityFee: %v", err)
+				}
+			} else {
+				for _, innerTxn := range txn.TxnMeta.(*AtomicTxnsWrapperMetadata).Txns {
```

Comment:
This is quite deeply nested, so cognitive complexity is pretty high in here. Can we write some sort of utility function that given an atomic transaction and a block reward recipient, iterates over it and return the total fees minus any fees paid by the block reward recipient? This function can return both BMF and old PoW fees. This way we can use the same function in each of the three places its used and not have such deeply nested for-loops and conditionals.

Reply:
Addressed with PR#1224.